### PR TITLE
gulp tasks

### PR DESCRIPTION
**Pull request by gertcuykens**
I know there are allot of things you are not going to like but lets discuss this because usually I had a good reason for it to make it work. :D

### CODE REVIEW

**Comment by gertcuykens** [comment]
Jesus how hard can it be to update my git with polymer that has merge conflicts... and then just add my changes again... its driving me nuts.

**Comment by addyosmani** [comment]
Please include a description with PRs :)

**Comment by gertcuykens** [comment]
1) Source maps
Ok lets start with the scss example to indicate how much I want for source maps to work like they do in the PR. Inspect my `Hello 2015!` ES6 example and then look at the style of it. Notice the exact line match to the original scss file :)

image

image

What ever the decision may be, make this work for developers, I am not talking about scss in particular, I am talking about source maps in general. Every single compiler depends on it! 

This comes with a price, that `.tmp` stuff you are doing... gone. `dist` is what I need.

2) Plugins 
Only google engineers understand the original `gulpfile.js` with all the vulcanising and linting going on. Fine but at least give us a tasks directorie like in the PR that allows us to easily configure that `gulpfile.js` beast. Don't make us edit that thing.

3) Reloading
image
 Add a workspace in chrome and edit anything you want HTML ES5 ES6 SCSS, it wil compile reload your changes on the fly. No it didn't do that in the original one, it only did it for html files.

Price rendering from dist only 
gulp serve => dist serving + watching changes
gulp serve:dist => just dist serving 

4) Testing
Yea yea yea I made sure that is still working too :p 
image

5) Babel
I vote to include it by default, in the PR its just a plugin task not hardcoded in the main `gulpfile.js` Developers kan just use ES5 or ES6. The old stuff and the new stuff works at the same time, you don't have to mess with anything it just works out of the box.

6) Your honor...
Polymer is advocating going forward pushing web components as far as they can, they even wrote there own shadow dom for crying out loud. Now do the same for javascript and make ES6 work by default and preferable scss too like in the PR :P Don't give me that vanilla bullshit please. https://github.com/google/material-design-lite has 15k stars and its uses scss hardcoded not even optional to build the thing and nobody is complaining about it. In the PR scss is just a plugin. It took me at least 48 hours to figure out how to make scss and source maps work with PSK. I understand the vanilla mind set but if somebody want chocolate sprinkles the need a PHD to understand how to apply a sprinkle function in PSK.

**Comment by gertcuykens** [comment]
Made the vulcanise process separate in serve:dist now.

**Comment by samccone** [comment]
great work on this stuff @gertcuykens I will take a look soonish :)

**Comment by gertcuykens** [comment]
Added 2015b example to make source maps work for webcomponents too.
Note you need to refresh the page while console is open before source the js source map is kicking in.

image

**Comment by gertcuykens** [comment]
Got rit of the history api middleware because it breaks browsing the components documentation. For example http://localhost:5000/bower_components/iron-input/

**Comment by samccone** [comment]
@gertcuykens mind doing a quick rebase on this one for me :)

**Comment by gertcuykens** [comment]
I merged everything into one commit but that didn't do what i want it to do :)
git reset --soft 89810334b352144f080774856e1cd11abacd3c68 
git commit -am "gulp tasks" --amend

**Comment by gertcuykens** [comment]
ok survived the rebase :)

**Comment by samccone** [comment]
Hey @gertcuykens ok I did a run through of this entire PR, there are a few things that I would like to resolve prior to this landing.

Can we just make this PR about moving the gulp tasks out of the single monolith file, we can address the other changes in this PR, like es6 support in other PRs. 

Thanks again and I am really excited to see this one land :airplane:

File: .jscsrc [hunk 1]
```diff
@@ -1,4 +1,5 @@
 {
+  "esnext": true,
```
Comment (samccone) [inline .jscsrc]: remove this plz :)

File: .jshintrc [hunk 2]
```diff
@@ -1,4 +1,5 @@
 {
+  "esnext": true,
```
Comment (samccone) [inline .jshintrc]: remove this plz :)

File: app/elements/elements.html [hunk 3]
```diff
@@ -39,3 +39,5 @@
 <link rel="import" href="../styles/shared-styles.html">
 <link rel="import" href="my-greeting/my-greeting.html">
 <link rel="import" href="my-list/my-list.html">
+<link rel="import" href="my-2015/my-2015.html">
```
Comment (samccone) [inline app/elements/elements.html]: remove these changes as well

File: app/elements/my-2015/my-2015.html [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  <template>
+    <style>
+    :host {
+      display: block;
+    }
+    </style>
+    <div>Hello 2015!</div>
+  </template>
+  <script src="my-2015.js"></script>
+</dom-module>
```
Comment (samccone) [inline app/elements/my-2015/my-2015.html]: :fire: this file

File: app/elements/my-2015/my-2015.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+class My2015 {
+  beforeRegister() {
+    this.is = 'my-2015';
+  }
+  ready() {
+    console.log('Hello 2015!');
+  }
+}
+Polymer(My2015);
```
Comment (samccone) [inline app/elements/my-2015/my-2015.js]: :fire: this file

File: app/elements/my-2015b/my-2015b.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+          display: block;
+        }
+      </style>
+      <div>Hello 2015b!</div>
+    `;
+    console.log('Hello 2015b!');
+  }
+}
+
+document.registerElement('my-2015b', My2015b);
```
Comment (samccone) [inline app/elements/my-2015b/my-2015b.js]: :fire: this file

File: app/index.html [hunk 7]
```diff
@@ -125,6 +125,9 @@
 
                 <p class="paper-font-subhead">You now have:</p>
                 <my-list></my-list>
+                <p class="paper-font-subhead">Including babel and scss plugins:</p>
```
Comment (samccone) [inline app/index.html]: :wave: remove please

File: gulpfile.js [hunk 8]
```diff
@@ -20,9 +20,10 @@ var merge = require('merge-stream');
 var path = require('path');
 var fs = require('fs');
 var glob = require('glob-all');
-var historyApiFallback = require('connect-history-api-fallback');
+//var historyApiFallback = require('connect-history-api-fallback');
```
Comment (samccone) [inline gulpfile.js]: undo this change

File: app/styles/main.scss [hunk 9]
```diff
@@ -7,6 +7,12 @@ Code distributed by Google as part of the polymer project is also
 subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
 */
 
+$primary-color: #3F51B5;
```
Comment (samccone) [inline app/styles/main.scss]: undo this change

File: gulpfile.js [hunk 10]
```diff
@@ -95,6 +85,25 @@ var optimizeHtmlTask = function(src, dest) {
...
+      rule: {
+        match: '<span id="browser-sync-binding"></span>',
+        fn: function(snippet) {
+          return snippet;
+        }
+      }
+    },
+    https: config.https,
+    server: config.dist(),
+    //middleware: [historyApiFallback()]
```
Comment (samccone) [inline gulpfile.js]: ? any reason why you commented this out?

File: gulpfile.js [hunk 11]
```diff
@@ -109,7 +118,7 @@ gulp.task('lint', function() {
   return gulp.src([
       'app/scripts/**/*.js',
       'app/elements/**/*.js',
-      'app/elements/**/*.html',
+      //'app/elements/**/*.html', ???warnings???
```
Comment (samccone) [inline gulpfile.js]: bring this back :)

File: gulpfile.js [hunk 12]
```diff
@@ -220,96 +229,68 @@ gulp.task('cache-config', function(callback) {
     if (error) {
       callback(error);
     } else {
-      config.precache = files;
+      files.push('index.html', './', 'bower_components/webcomponentsjs/webcomponents-lite.min.js');
```
Comment (samccone) [inline gulpfile.js]: looks like we might have messed up the rebase here, this should be using glob2

Comment (samccone) [inline gulpfile.js]: look on master to see how it is currently done

File: gulpfile.js [hunk 13]
```diff
@@ -220,96 +229,68 @@ gulp.task('cache-config', function(callback) {
...
+      fs.writeFile(configPath, JSON.stringify(cache), callback);
     }
   });
 });
 
 // Clean output directory
-gulp.task('clean', function() {
-  return del(['.tmp', 'dist']);
+gulp.task('clean', function(cb) {
+  del(['dist'], cb);
```
Comment (samccone) [inline gulpfile.js]: return and now CB here (look how we are doing it on master)

File: package.json [hunk 14]
```diff
@@ -25,6 +25,10 @@
     "gulp-uglify": "^1.2.0",
     "gulp-useref": "^2.1.0",
     "gulp-vulcanize": "^6.0.0",
+    "gulp-sass": "^2.0.*",
```
Comment (samccone) [inline package.json]: :x:

File: jsconfig.json [hunk 15]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (samccone) [inline jsconfig.json]: :x: please

File: package.json [hunk 16]
```diff
@@ -25,6 +25,10 @@
...
+    "gulp-sourcemaps": "^1.3.0",
+    "gulp-babel": "^5.2.1",
```
Comment (samccone) [inline package.json]: :x:

File: tasks/babel.js [hunk 17]
```diff
@@ -0,0 +1,11 @@
+var gulp = require('gulp');
```
Comment (samccone) [inline tasks/babel.js]: remove this task, instead we should update the recipe doc for this PR with how to add this

File: tasks/scss.js [hunk 18]
```diff
@@ -0,0 +1,13 @@
+var gulp = require('gulp');
```
Comment (samccone) [inline tasks/scss.js]: remove this task, instead we should update the recipe doc for this PR with how to add this

File: tasks/ts.js [hunk 19]
```diff
@@ -0,0 +1,19 @@
+var gulp = require('gulp');
```
Comment (samccone) [inline tasks/ts.js]: remove this task, instead we should update the recipe doc for this PR with how to add this

File: tsconfig.json [hunk 20]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (samccone) [inline tsconfig.json]: :x:

File: gulpfile.js (hunk 8 above)
Comment (gertcuykens) [inline gulpfile.js]: historyApiFallback breaks the bower_components browser for example try http://localhost:5000/bower_components/iron-input/

File: gulpfile.js (hunk 10 above)
Comment (gertcuykens) [inline gulpfile.js]: historyApiFallback breaks the bower_components browser for example try http://localhost:5000/bower_components/iron-input/

**Comment by gertcuykens** [comment]
@samccone did you get the historyApiFallback comment?
It breaks the bower_components browser for example try http://localhost:5000/bower_components/iron-input/

File: gulpfile.js (hunk 10 above)
Comment (samccone) [inline gulpfile.js]: hmm if this is the case, mind opening a new standalone PR so we can discuss it there...

thanks!

**Comment by gertcuykens** [comment]
Went trough the list but i can't find why i broke the copy task now, it worked in my commit before this one :)

**Comment by gertcuykens** [comment]
ok fixed it but still needs a other round of code review because there are some dogy things that need to be addressed :P for example the gulp cleaning process

**Comment by samccone** [comment]
hey @gertcuykens this is looking really good, mind doing 1 more round of rebasing :)

I see this landing this week hopefully!

**Comment by samccone** [comment]
@chuckh would you mind looking this over and perhaps call out the recipes that need to be updated?

**Comment by chuckh** [comment]
@samccone will do today. Looks like this also needs a rebase.

**Comment by gertcuykens** [comment]
Again! I am on it :)

File: gulpfile.js [hunk 21]
```diff
@@ -37,21 +38,14 @@ var AUTOPREFIXER_BROWSERS = [
   'bb >= 10'
 ];
 
-var DIST = 'dist';
-
```
Comment (chuckh) [inline gulpfile.js]: Where do you set this now if you want it different from `dist`?

File: tasks/default.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+  https: false,
+  watch: function(){
+    gulp.watch(['app/**/*.html'], [reload]);
+    gulp.watch(['app/{scripts,elements}/**/{*.js,*.html}'], ['jshint', reload]);
+    gulp.watch(['app/images/**/*'], reload);
+  }
+};
+
+// The default task
+gulp.task('default', function(){ //['clean'],
```
Comment (gertcuykens) [inline tasks/default.js]: need explanation why I can't do `gulp.task('default', ['clean'], function(){`anymore here? I clean at the same time it copies ???

File: tasks/default.js [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+  // Uncomment 'cache-config' if you are going to use service workers.
+  runSequence(
+    ['copy', 'styles'],
+    'elements',
+    ['lint', 'images', 'fonts', 'html']
```
Comment (gertcuykens) [inline tasks/default.js]: I want to get rit of 'html' and put it in the vulcanise task but if I do that browsersync messes up?

File: gulpfile.js (hunk 21 above)
Comment (gertcuykens) [inline gulpfile.js]: all config stuf is in one file now `tasks/default.js`

**Comment by chuckh** [comment]
@samccone looks like README.md and add-es2015-support-babel.md need to be updated. See below.

> Price rendering from dist only 
gulp serve => dist serving + watching changes
gulp serve:dist => dist vulcanise + dist serving 
gulp => becomes gulp dist - *

Need to change https://github.com/PolymerElements/polymer-starter-kit/blob/master/README.md
1. For `gulp dist` need to change at https://github.com/PolymerElements/polymer-starter-kit#build--vulcanize 
2. Should add section for `gulp serve:dist` under https://github.com/PolymerElements/polymer-starter-kit#development-workflow
3. Need to update under https://github.com/PolymerElements/polymer-starter-kit#enable-service-worker-support item 3 for these changes
4. Need to update https://github.com/PolymerElements/polymer-starter-kit#disable-service-worker-support-after-you-enabled-it item 3

Need to change https://github.com/PolymerElements/polymer-starter-kit/blob/master/docs/add-es2015-support-babel.md
1 Need to update https://github.com/PolymerElements/polymer-starter-kit/blob/master/docs/add-es2015-support-babel.md#integrating-the-transpile-task
2. Should review complete doc file and update

File: gulpfile.js [hunk 24]
```diff
@@ -221,91 +231,53 @@ gulp.task('cache-config', function(callback) {
...
-      }
-    },
-    // Run as an https by uncommenting 'https: true'
-    // Note: this uses an unsigned certificate which on first access
-    //       will present a certificate warning in the browser.
-    // https: true,
-    server: dist(),
-    middleware: [historyApiFallback()]
-  });
+// Clean dist directory
```
Comment (gertcuykens) [inline gulpfile.js]: need expert advice how to clean the dist up properly after vulcanising the lines below are just a starting point

File: gulpfile.js [hunk 25]
```diff
@@ -221,91 +231,53 @@ gulp.task('cache-config', function(callback) {
...
-  // Uncomment 'cache-config' if you are going to use service workers.
+// Build production files
+gulp.task('dist', ['default'], function() {
   runSequence(
-    ['copy', 'styles'],
-    'elements',
-    ['lint', 'images', 'fonts', 'html'],
-    'vulcanize', // 'cache-config',
-    cb);
+    //'html',
```
Comment (gertcuykens) [inline gulpfile.js]: i want the html thing to run here but first i need to know why browsersync is messing default up?

File: tasks/default.js [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+    //['copy'],
+    //['lint', 'images', 'fonts']
+    // 'cache-config'
+  );
+});
```
Comment (gertcuykens) [inline tasks/default.js]: final note, I don't know why but babel and sccs stuf I added first fixed the cleaning and html problem?

**Comment by gertcuykens** [comment]
I will edit the ES6 doc after the other issues are fixed. basically just adding a tasks file like this
```
var gulp = require('gulp');
var $ = require('gulp-load-plugins')();

gulp.task('babel', function () {
  return gulp.src(['app/**/*.{js,html}'])
    .pipe($.sourcemaps.init())
    .pipe($.if('*.html', $.crisper())) // Extract JS from .html files
    .pipe($.if('*.js', $.babel()))
    .pipe($.sourcemaps.write('.')) // {sourceMappingURL: function(file) {return '/' + file.relative + '.map';}}
    .pipe(gulp.dest('dist/'));
});
```

**Comment by robdodson** [comment]
Per @samccone's comment. I think it's a good idea to break apart the gulpfile into smaller pieces but we probably have to do this in baby steps. I think trying to add in source maps, es6, scss at the same time is too much all at once. It may be better to start a new PR that just attempts to break things into individual tasks.

**Comment by gertcuykens** [comment]
@robdodson yep got rid of it all, in previous rebase :) but i see i need to rebase again to make it uptodate with master again.

**Comment by gertcuykens** [comment]
Rebase done, I need a gulp expert to look at it and explain why gulp is making a mess of it, I can fix it with trial and error by moving some commands around but I need explanation.

**Comment by addyosmani** [comment]
Thanks! Side note: If we do want to land this its a large enough change that it will need to wait until 1.3 imo. There are quite a few other fixes we still need to get into master go bring 1.2 back to stability.

**Comment by gertcuykens** [comment]
No problem, if you can just try it and explain the gulp behavior that would be wonderful :)

File: gulpfile.js [hunk 27]
```diff
@@ -96,6 +86,25 @@ var optimizeHtmlTask = function(src, dest) {
...
+      rule: {
+        match: '<span id="browser-sync-binding"></span>',
+        fn: function(snippet) {
+          return snippet;
+        }
+      }
+    },
+    https: config.https,
+    server: config.dist(),
+    middleware: [historyApiFallback()]
```
Comment (gertcuykens) [inline gulpfile.js]: historyApiFallback() breaks browsing the components documentation. For example http://localhost:5000/bower_components/iron-input/

File: tasks/default.js [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+  //       will present a certificate warning in the browser.
+  https: false,
+  watch: function(){
+    gulp.watch(['app/**/*.html'], [reload]);
+    gulp.watch(['app/{scripts,elements}/**/{*.js,*.html}'], ['jshint', reload]);
+    gulp.watch(['app/images/**/*'], reload);
+  }
+};
+
+// The default task
```
Comment (gertcuykens) [inline tasks/default.js]: Why is gulp is making a mess of the default task???

File: tasks/default.js [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+gulp.task('default', ['clean'], function(){
+  // Uncomment 'cache-config' if you are going to use service workers.
+  runSequence(
+    ['copy', 'styles'],
+    'elements',
+    ['lint', 'images', 'fonts']
+    // 'cache-config'
+  );
+});
```
Comment (gertcuykens) [inline tasks/default.js]: Why is gulp making a mess of the default task above???

File: gulpfile.js [hunk 30]
```diff
@@ -220,91 +230,53 @@ gulp.task('cache-config', function(callback) {
...
+gulp.task('vulclean', function() {
+  return del([
+    //'dist/bower_components/**/*',
+    //'!dist/bower_components/webcomponentsjs',
+    'dist/**/*.map',
+    'dist/elements/**/*.html',
+    '!dist/elements/elements.vulcanized.html',
+    '!dist/elements/routing.html'
+  ]);
 });
```
Comment (gertcuykens) [inline gulpfile.js]: Need expert advice how to clean the dist up properly after vulcanising the lines above are just a starting point.

Comment (robdodson) [inline gulpfile.js]: I'm working on a PR here which simplifies the way we build and vulcanize. https://github.com/PolymerElements/polymer-starter-kit/pull/568

You should be able to just keep the current `clean` task and not need `vulclean`

File: gulpfile.js [hunk 31]
```diff
@@ -23,6 +23,7 @@ var glob = require('glob-all');
 var historyApiFallback = require('connect-history-api-fallback');
 var packageJson = require('./package.json');
 var crypto = require('crypto');
+var config = require('./tasks/default.js');
```
Comment (robdodson) [inline gulpfile.js]: you could also do `var dist = config.dist` so you can just write `dist()` instead of `config.dist()` everywhere. Would be much cleaner.

Comment (gertcuykens) [inline gulpfile.js]: nice! :D

File: gulpfile.js (hunk 30 above)
Comment (gertcuykens) [inline gulpfile.js]: OK got it wil remove

**Comment by robdodson** [comment]
I'd like to land https://github.com/PolymerElements/polymer-starter-kit/pull/568 before moving forward with this. @addyosmani @samccone @chuckh would either of you have time to review that PR?

**Comment by addyosmani** [comment]
I think we just need to get the recipes tackled and 568 can land.

**Comment by gertcuykens** [comment]
Can I suggest a short cut, instead of putting them in the repository put them in the github wiki section so they don't block development? Also once we do this commit they all need to be changed again.

File: gulpfile.js (hunk 31 above)
Comment (gertcuykens) [inline gulpfile.js]: i will change this when everything else works because i use it as a names space to figure out what i messing up by searching for the word 'config' in the gulpfile :)

**Comment by gertcuykens** [comment]
rebased

**Comment by samccone** [comment]
oh no :( 1 more time ha

**Comment by gertcuykens** [comment]
it lasted 5 seconds lol

**Comment by gertcuykens** [comment]
Rebased, also tried to stick to the readme commands documentation.

**Comment by gertcuykens** [comment]
So question number one too expert, why is this happening when i type 3 times gulp :D

$ gulp
[22:35:43] Using gulpfile ~/Desktop/polymer-starter-kit/gulpfile.js
[22:35:43] Starting 'clean'...
[22:35:43] Finished 'clean' after 2.81 ms
[22:35:43] Starting 'default'...
[22:35:43] Starting 'copy'...
[22:35:43] Starting 'styles'...
stream.js:94
      throw er; // Unhandled stream error in pipe.
            ^
Error: ENOENT, stat '/Users/gert/Desktop/polymer-starter-kit/dist/styles/main.css'
    at Error (native)

$ gulp
[22:36:10] Using gulpfile ~/Desktop/polymer-starter-kit/gulpfile.js
[22:36:10] Starting 'clean'...
[22:36:10] Finished 'clean' after 3.13 ms
[22:36:10] Starting 'default'...
[22:36:10] Starting 'copy'...
[22:36:10] Starting 'styles'...
[22:36:11] styles all files 98 B
[22:36:11] Finished 'styles' after 543 ms
[22:36:11] copy all files 1.04 MB
[22:36:11] Finished 'copy' after 695 ms
[22:36:11] Starting 'elements'...
[22:36:11] Finished 'elements' after 5.16 ms
[22:36:11] Starting 'ensureFiles'...
[22:36:11] Finished 'ensureFiles' after 203 μs
[22:36:11] Starting 'lint'...
[22:36:12] Starting 'images'...
[22:36:13] Starting 'fonts'...
[22:36:13] Starting 'html'...
[22:36:13] Finished 'fonts' after 317 ms
[22:36:15] Finished 'lint' after 3.99 s
[22:36:15] html all files 91.61 kB
[22:36:15] Finished 'html' after 2.4 s
[22:36:15] gulp-imagemin: Minified 6 images (saved 494 B - 1.4%)
[22:36:15] images all files 35.41 kB
[22:36:15] Finished 'images' after 2.7 s
[22:36:15] Starting 'vulcanize'...
[22:36:16] vulcanize all files 482.06 kB
[22:36:16] Finished 'vulcanize' after 946 ms
[22:36:16] Finished 'default' after 5.73 s

$ gulp
[22:36:21] Using gulpfile ~/Desktop/polymer-starter-kit/gulpfile.js
[22:36:21] Starting 'clean'...
[22:36:21] Finished 'clean' after 2.72 ms
[22:36:21] Starting 'default'...
[22:36:21] Starting 'copy'...
[22:36:21] Starting 'styles'...
stream.js:94
      throw er; // Unhandled stream error in pipe.
            ^
Error: ENOENT, chmod '/Users/gert/Desktop/polymer-starter-kit/dist/bower_components/webcomponentsjs/CustomElements.js'
    at Error (native)

**Comment by gertcuykens** [comment]
Everything is working now except gulp test

`null is not an object (evaluating 'header.textContent')`

**Comment by abdonrd** [comment]
Closed in favor to [PolymerElements/generator-polymer-init-custom-build](https://github.com/PolymerElements/generator-polymer-init-custom-build).
